File: backend/src/main/java/com/bot/service/user/UserService.java
package com.bot.service.user;

import com.bot.controller.user.dto.ChangePasswordRequest;
import com.bot.controller.user.dto.ProfileUpdateRequest;
import com.bot.entity.User;
import com.bot.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.UUID;

/**
 * Service xử lý các nghiệp vụ liên quan đến User (người dùng).
 * Bao gồm: xem hồ sơ, cập nhật thông tin, đổi mật khẩu, cập nhật avatar.
 */
@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    // NOTE: Nghiệp vụ đăng ký nằm ở AuthService.
    // Hàm createUser này chủ yếu dùng cho admin.
    public User createUser(User user) {
        if (userRepository.existsByUsername(user.getUsername())) {
            throw new IllegalArgumentException("Username đã được sử dụng");
        }
        return userRepository.save(user);
    }

    /**
     * Lấy thông tin user hiện tại đang đăng nhập dựa vào JWT token.
     * Sử dụng SecurityContextHolder để lấy username từ principal.
     *
     * @return Optional<User> chứa thông tin user đang đăng nhập
     */
    public Optional<User> getCurrentUser() {
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        String username;
        if (principal instanceof UserDetails userDetails) {
            username = userDetails.getUsername();
        } else {
            username = principal.toString();
        }
        return userRepository.findByUsername(username);
    }

    /**
     * Tìm kiếm thông tin user theo username.
     */
    public Optional<User> findByUsername(String username) {
        return userRepository.findByUsername(username);
    }

    /**
     * Tìm kiếm thông tin user theo UUID.
     */
    public Optional<User> findById(UUID id) {
        return userRepository.findById(id);
    }

    /**
     * Cập nhật thông tin hồ sơ cá nhân (fullName, username, email).
     *
     * @param id  UUID của user cần cập nhật
     * @param req DTO chứa thông tin mới
     * @return User đã được cập nhật
     * @throws IllegalArgumentException nếu user không tồn tại hoặc username/email đã bị chiếm
     */
    public User updateProfile(UUID id, ProfileUpdateRequest req) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Không tìm thấy người dùng với ID: " + id));

        // Update fullName nếu được cung cấp
        if (req.getFullName() != null && !req.getFullName().isBlank()) {
            user.setFullName(req.getFullName());
        }

        // Update username nếu được cung cấp và không trùng
        if (req.getUsername() != null && !req.getUsername().isBlank()) {
            if (!req.getUsername().equals(user.getUsername())
                    && userRepository.existsByUsername(req.getUsername())) {
                throw new IllegalArgumentException("Tên đăng nhập '" + req.getUsername() + "' đã được sử dụng.");
            }
            user.setUsername(req.getUsername());
        }

        // Update email nếu được cung cấp và không trùng
        if (req.getEmail() != null && !req.getEmail().isBlank()) {
            if (!req.getEmail().equals(user.getEmail())
                    && userRepository.existsByEmail(req.getEmail())) {
                throw new IllegalArgumentException("Email '" + req.getEmail() + "' đã được sử dụng.");
            }
            user.setEmail(req.getEmail());
        }

        return userRepository.save(user);
    }

    /**
     * Đổi mật khẩu người dùng.
     * Yêu cầu xác minh mật khẩu cũ và kiểm tra khớp mật khẩu mới.
     *
     * @param id  UUID của user
     * @param req DTO chứa mật khẩu hiện tại và mật khẩu mới
     * @throws IllegalArgumentException nếu mật khẩu cũ sai hoặc mật khẩu mới không khớp
     */
    public void changePassword(UUID id, ChangePasswordRequest req) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Không tìm thấy người dùng với ID: " + id));

        // Xác minh mật khẩu hiện tại
        if (!passwordEncoder.matches(req.getCurrentPassword(), user.getPasswordHash())) {
            throw new IllegalArgumentException("Mật khẩu hiện tại không chính xác.");
        }

        // Kiểm tra mật khẩu mới và xác nhận phải khớp nhau
        if (!req.getNewPassword().equals(req.getConfirmPassword())) {
            throw new IllegalArgumentException("Mật khẩu mới và xác nhận mật khẩu không khớp.");
        }

        // Hash và lưu mật khẩu mới
        user.setPasswordHash(passwordEncoder.encode(req.getNewPassword()));
        userRepository.save(user);
    }

    /**
     * Cập nhật đường dẫn URL ảnh avatar cho người dùng.
     *
     * @param id        UUID của user
     * @param avatarUrl Đường dẫn URL tĩnh của ảnh vừa upload
     * @return User đã được cập nhật
     */
    public User updateAvatarUrl(UUID id, String avatarUrl) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Không tìm thấy người dùng với ID: " + id));
        user.setAvatarUrl(avatarUrl);
        return userRepository.save(user);
    }

    /**
     * Cập nhật sở thích (preferences) cho người dùng.
     *
     * @param id              UUID của user
     * @param preferencesJson Chuỗi JSON preferences
     * @return User đã được cập nhật
     */
    public User updatePreferences(UUID id, String preferencesJson) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Không tìm thấy người dùng"));
        user.setPreferences(preferencesJson);
        return userRepository.save(user);
    }

    /**
     * Xóa ảnh avatar của người dùng.
     *
     * @param id UUID của user
     * @return User đã được cập nhật (avatarUrl = null)
     */
    public User removeAvatar(UUID id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Không tìm thấy người dùng với ID: " + id));
        user.setAvatarUrl(null);
        return userRepository.save(user);
    }
}

File: frontend/src/lib/userService.js
/**
 * userService.js — Dịch vụ API quản lý thông tin người dùng.
 *
 * Module này chứa các hàm gọi API liên quan đến:
 * - GET /me: Lấy thông tin user đang đăng nhập (dùng JWT)
 * - GET /{id}: Lấy thông tin user theo UUID
 * - PUT /{id}/profile: Cập nhật thông tin hồ sơ (fullName, username, email)
 * - PUT /{id}/password: Đổi mật khẩu (cần xác minh mật khẩu cũ)
 * - POST /{id}/avatar: Upload ảnh đại diện từ máy tính lên server
 * - PUT /{id}/preferences: Cập nhật sở thích cá nhân
 */
import api from './api';

const userService = {
  /**
   * Lấy thông tin user đang đăng nhập dựa vào JWT token.
   * Token được tự động gửi kèm qua interceptor trong api.js.
   *
   * GET /api/v1/users/me
   * @returns {Promise<Object>} UserProfileResponse { id, username, email, fullName, avatarUrl, role, createdAt }
   */
  async getCurrentUser() {
    const response = await api.get('/v1/users/me');
    return response.data;
  },

  /**
   * Lấy thông tin user theo UUID.
   *
   * GET /api/v1/users/{id}
   * @param {string} id - UUID của người dùng
   * @returns {Promise<Object>} UserProfileResponse
   */
  async getUserById(id) {
    const response = await api.get(`/v1/users/${id}`);
    return response.data;
  },

  /**
   * Cập nhật thông tin hồ sơ cá nhân.
   *
   * PUT /api/v1/users/{id}/profile
   * Body: { fullName, username, email } — các trường không bắt buộc
   *
   * @param {string} id - UUID người dùng
   * @param {Object} data - { fullName?, username?, email? }
   * @returns {Promise<Object>} UserProfileResponse đã cập nhật
   */
  async updateProfile(id, data) {
    const response = await api.put(`/v1/users/${id}/profile`, data);
    return response.data;
  },

  /**
   * Đổi mật khẩu người dùng.
   * Backend sẽ xác minh mật khẩu cũ trước khi đổi.
   *
   * PUT /api/v1/users/{id}/password
   * Body: { currentPassword, newPassword, confirmPassword }
   *
   * @param {string} id - UUID người dùng
   * @param {Object} data - { currentPassword, newPassword, confirmPassword }
   * @returns {Promise<Object>} { message: "Đổi mật khẩu thành công." }
   * @throws {AxiosError} 400 nếu sai mật khẩu cũ hoặc mật khẩu mới không khớp
   */
  async changePassword(id, data) {
    const response = await api.put(`/v1/users/${id}/password`, data);
    return response.data;
  },

  /**
   * Upload ảnh đại diện (avatar) từ máy tính lên server.
   * Sử dụng FormData để gửi file dạng multipart/form-data.
   * Server lưu file vào src/main/resources/db/uploads/ và trả về URL.
   *
   * POST /api/v1/users/{id}/avatar
   *
   * @param {string} id - UUID người dùng
   * @param {File} file - File ảnh từ input[type="file"]
   * @returns {Promise<Object>} UserProfileResponse với avatarUrl mới
   */
  async uploadAvatar(id, file) {
    const formData = new FormData();
    formData.append('file', file);
    const response = await api.post(`/v1/users/${id}/avatar`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return response.data;
  },

  /**
   * Xóa ảnh đại diện (avatar) của người dùng.
   *
   * @param {string} id - UUID người dùng
   * @returns {Promise<Object>} UserProfileResponse với avatarUrl = null
   */
  async removeAvatar(id) {
    const response = await api.delete(`/v1/users/${id}/avatar`);
    return response.data;
  },

  /**
   * Cập nhật sở thích cá nhân hóa (preferences).
   *
   * PUT /api/v1/users/{id}/preferences
   *
   * @param {string} id - UUID người dùng
   * @param {Object} preferences - Object { language, theme, notifications, ... }
   * @returns {Promise<Object>} UserProfileResponse đã cập nhật
   */
  async updatePreferences(id, preferences) {
    const response = await api.put(`/v1/users/${id}/preferences`, preferences);
    return response.data;
  },

  /**
   * Tạo người dùng mới — CHỈ DÀNH CHO ADMIN.
   *
   * POST /api/v1/users
   *
   * @param {Object} userData - { username, email, password, role }
   * @returns {Promise<Object>} UserProfileResponse mới tạo
   */
  async createUser(userData) {
    const response = await api.post('/v1/users', userData);
    return response.data;
  },
};

export default userService;
